Sources/HTMLString/HTMLString.swift

```diff
@@ -140,35 +144,70 @@ extension String {
     /// | `&` | `&` | Not an entity |
     ///
 
-    public mutating func removeHTMLEntities() {
-        var searchSubstring = self[startIndex ..< endIndex]
+    public func removeHTMLEntities() -> String {
+        var result = ""
+        var currentIndex = startIndex
 
-        while let delimiterIndex = searchSubstring.firstIndex(of: "&") {
+        while let delimiterIndex = self[currentIndex...].firstIndex(of: "&") {
             // Avoid unnecessary operations
-            guard let semicolonIndex = searchSubstring[delimiterIndex ..< endIndex].firstIndex(of: ";") else {
-                break
-            }
+            var semicolonIndex = self.index(after: delimiterIndex)
 
             // Parse the last sequence (ex: Fish & chips &amp; sauce -> "&amp;" instead of "& chips &amp;")
-            let lastDelimiterIndex = searchSubstring[delimiterIndex ..< semicolonIndex].lastIndex(of: "&") ?? delimiterIndex
+            var lastDelimiterIndex = delimiterIndex
+
+            while semicolonIndex != endIndex, self[semicolonIndex] != ";" {
+                if self[semicolonIndex] == "&" {
+                    lastDelimiterIndex = semicolonIndex
+                }
+
+                semicolonIndex = self.index(after: semicolonIndex)
+            }
+
+            // Fast path if semicolon doesn't exists in current range
+            if semicolonIndex == endIndex {
+                result.append(contentsOf: self[currentIndex..<semicolonIndex])
+                return result
+            }
```

Comment:
What is the benefit of updating the lookup logic here? Would it be possible to use the same logic as before, which seems to be more readable for me?

```swift
guard let semicolonIndex = self[delimiterIndex ..< endIndex].firstIndex(of: ";") else {
    continue
}		
 
// Parse the last sequence (ex: Fish & chips &amp; sauce -> "&amp;" instead of "& chips &amp;")
let lastDelimiterIndex = self[delimiterIndex ..< semicolonIndex].lastIndex(of: "&") ?? delimiterIndex
```

Reply:
I wanted to traverse string only one time. Your code is traversing (delimiterIndex..<semicolonIndex)range two times.(One is left to right, the other is right-to-left)

Reply:
Ah right, good point